### PR TITLE
CB-12774: Support scoped plugins

### PR DESCRIPTION
**Pull request by oliversalzburg**
### Platforms affected
All

### What does this PR do?
See #602 for details.

### Checklist
- [x] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [x] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [x] Added automated test coverage as appropriate for this change.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{
```
Comment (dpogue) [inline tsconfig.json]: We probably don't want to add a tsconfig to the repo without further discussion?

Comment (raphinesse) [reply tsconfig.json]: It's from a commit labeled TEMP. I guess the original author used it to aid the refactoring.

**Comment by oliversalzburg** [comment]
I also dropped 6acff9309641c080dfd9227161384f027a013496 as I think it should have been removed. Not so sure about the 3 failures. I didn't see those earlier. I'll look into them. If someone has any ideas beforehand, that'd be great.

**Comment by raphinesse** [comment]
@oliversalzburg To me 6acff93 looked like some change to a class from `cordova-common` that was needed for this change and that should have been ported to common once this was finished.

**Comment by oliversalzburg** [comment]
@raphinesse I think you're right. I'm debugging the failing tests right now and it appears the `getAllHelper` is broken. What would you suggest? Bringing the change back or putting the change into common?

**Comment by raphinesse** [comment]
I'd say leave it here until we're certain we have what we want. Then move it to common

**Comment by oliversalzburg** [comment]
I think I had that timeout-related failure locally as well. I'll look into it.

**Comment by oliversalzburg** [comment]
No idea about that AppVeyor failure right now. I'll head to bed soon and then to Amsterdam over the weekend. I can look into it next week hopefully. I'm open for ideas of course :D

**Comment by raphinesse** [comment]
So with the tests passing, do you think this is ready for a more thorough review @oliversalzburg? Or do you think there is still work to be done before that? (Apart from the changes that have to be moved to common if the rest checks out)

**Comment by oliversalzburg** [comment]
I'd say get it into master ASAP. Then people could test with nightly builds, that's what I did.

But because it didn't work out, but I had to move forward, all our packages are now published without a scope and we've abandoned npm (the company/registry) altogether. So I currently have neither a way of testing the new use case nor do I have any interest in moving back.

It would probably make sense to publish cordova under an npm scope at some point in the future. That would make for enough testing.

**Comment by raphinesse** [comment]
> I'd say get it into master ASAP. Then people could test with nightly builds, that's what I did.

While crowd-sourcing is a fine thing, I still have to review and understand it, before I approve it :wink: 

> It would probably make sense to publish cordova under an npm scope at some point in the future. That would make for enough testing.

Yeah, that's been on my silent wish list for some time now. But there's tons of more important stuff to do than that.

Do you have any objections against me rebasing and squashing these commits as is necessary to further clean up this PR?

**Comment by oliversalzburg** [comment]
> Do you have any objections against me rebasing and squashing these commits as is necessary to further clean up this PR?

None at all. Do as you see fit :)

**Comment by oliversalzburg** [comment]
The build failures on AppVeyor could be due to https://github.com/apache/cordova-common/issues/64

**Comment by erisu** [comment]
@oliversalzburg Just looking at the error, it feel more like a path issue. From the output, somewhere we are not building the path with the correct separator.

```
Expected spy copySync to have been called with [ 'C:\projects\cordova-lib\spec\plugman\plugins\org.test.plugins.childbrowser', 'C:\Users\appveyor\AppData\Local\Temp\1\plugman\fetch\org.test.plugins.childbrowser', <jasmine.objectContaining(Object({ dereference: true }))> ] but actual calls were [ 'C:/projects/cordova-lib/spec/plugman/plugins/org.test.plugins.childbrowser', 'C:\Users\appveyor\AppData\Local\Temp\1\plugman\fetch\org.test.plugins.childbrowser', Object({ dereference: true, clobber: true, overwrite: true }) ].
```

The first arg expected
  `C:\projects\cordova-lib\spec\plugman\plugins\org.test.plugins.childbrowser`
But has
  `C:/projects/cordova-lib/spec/plugman/plugins/org.test.plugins.childbrowser`.

Revered separator...

**Comment by oliversalzburg** [comment]
@erisu Good point. Didn't notice that. I assume the `PluginInfoProvider` always returns a path with forward slashes?

**Comment by raphinesse** [comment]
I did take a quick look last night, but couldn't figure out what went wrong or why it only started to occur after my last changes.

I'll be on vacation for some time now, so unfortunately I'll not be able to figure it out myself.

**Comment by oliversalzburg** [comment]
I was wondering if it comes from here https://github.com/apache/cordova-lib/blob/master/src/plugman/fetch.js#L282 because it uses "some object" and just pulls whatever path is stored on it. I'm seeing `path.normalize` all over the place, probably to satisfy the tests, as this should have no real relevance in practice. Maybe just sprinkle a bit of them into the codebase? Or maybe just into the test?

**Comment by raphinesse** [comment]
I agree that the test should probably normalize paths. Except if we expect a normalized form for some specific reason

**Comment by oliversalzburg** [comment]
Build is green now. Would be nice if we can still get this into 9.0

**Comment by raphinesse** [comment]
The problem with getting this in 9.0 is the change that actually has to be made in `cordova-common`. I don't think we want to go back and release a new major of common right now.

We _could_ leave the overriding of the functionality from `common` in here to fast-track this change into 9.0. But if we do so, we should probably prefer a noninvasive specialization of the class from `common` over monkey-patching it. Plus, we should have a to-do issue to finally move the change to `common` so this is not forgotten.

Jus my 2¢

**Comment by dpogue** [comment]
Is the change in cordova-common a breaking change?

I thought it was just an addition, but I now realize I misread the diff (of course it's just an addition here, because we're redefining it)

**Comment by raphinesse** [comment]
@dpogue We are redefining an existing method and changing its behavior. AFAICT that unfortunately necessitates a major release.

**Comment by raphinesse** [comment]
In reply to @purplecabbage's comment from the mailing list regarding this being only a minor change: Frankly I'm just not sure about it. Before this change, scopes have been deemed part of the plugin version by our code. And I faintly seem to remember @dpogue telling someone that this served the purpose of being able to override existing plugins with forked versions. I don't know if that is still an officially supported use case, but if it is I would assume we are breaking it with this change.

**Comment by dpogue** [comment]
To avoid this falling off the radar again, can someone please open a pull request on cordova-common with the needed changes and label it for next major?

**Comment by raphinesse** [comment]
@dpogue Since I think you once said that you were involved in the current handling of scoped plugins I would be interested if _you_ think that these are actually breaking changes? Is the _scoped plugins to override other plugins_ functionality still functional in Cordova 9? Is it actually publicly documented somewhere?

Because it would be really cool to be able to release this in a minor or patch release.

**Comment by l3ender** [comment]
Just ran into this issue in our environment--any chance of the fix getting merged/released shortly? Thanks all for the work on it!

**Comment by raphinesse** [comment]
@l3ender I am currently working on integrating it. I hope to be done with it this week. But it will only be officially released in the next major release (Cordova 10). That is scheduled for the beginning of 2020 IIRC.

However, as soon as the changes are merged, they can be tested in `cordova@nightly`. It would be great to get some early feedback on this feature that way.

**Comment by raphinesse** [comment]
See #821 for an updated version of this PR

**Comment by raphinesse** [comment]
To all interested parties: #821 has been merged and will be available for testing in the next `cordova@nightly`

**Pull request closed without merging**